hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/lock/FileSystemBasedLockProvider.java

```diff
@@ -0,0 +1,97 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.client.lock;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hudi.common.config.LockConfiguration;
+import org.apache.hudi.common.lock.LockProvider;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.exception.HoodieLockException;
+
+import java.io.IOException;
+
+import static org.apache.hudi.common.config.LockConfiguration.FILESYSTEM_LOCK_PATH_PROP;
+import static org.apache.hudi.common.config.LockConfiguration.HOODIE_LOCK_ACQUIRE_NUM_RETRIES_PROP;
+import static org.apache.hudi.common.config.LockConfiguration.HOODIE_LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP;
+
+/**
+ * This lock provider is used to testing purposes only. It provides a simple file system based lock using HDFS atomic
+ * create operation. This lock does not support cleaning/expiring the lock after a failed write hence cannot be used
+ * in production environments.
+ */
+public class FileSystemBasedLockProvider extends LockProvider {
+
+  private static final String LOCK_NAME = "acquired";
+
+  private String lockPath;
+  private FileSystem fs;
+
+  public FileSystemBasedLockProvider(LockConfiguration lockConfiguration, final Configuration configuration) {
+    try {
+      this.lockConfiguration = lockConfiguration;
+      this.lockPath = lockConfiguration.getConfig().getString(FILESYSTEM_LOCK_PATH_PROP);
+      this.fs = FileSystem.get(configuration);
+    } catch (IOException io) {
+      throw new HoodieIOException("Unable to create file systems", io);
+    }
+  }
+
+  @Override
+  public void acquireLock() {
+    try {
+      fs.create(new Path(lockPath + "/" + LOCK_NAME)).close();
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to acquire lock", e);
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    fs.close();
+  }
+
+  @Override
+  public boolean tryLock() {
+    try {
+      int numRetries = 0;
+      while (fs.exists(new Path(lockPath + "/" + LOCK_NAME))
+          && (numRetries <= lockConfiguration.getConfig().getInteger(HOODIE_LOCK_ACQUIRE_NUM_RETRIES_PROP))) {
+        Thread.sleep(lockConfiguration.getConfig().getInteger(HOODIE_LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP));
+      }
+      acquireLock();
+      return true;
+    } catch (IOException | InterruptedException e) {
+      throw new HoodieLockException("Failed to acquire lock", e);
+    }
+  }
+
+  @Override
+  public void unlock() {
```

Comment:
whats the guarantee that `unlock()` will fail for thread B, if thread A had actually created the lock file. Don't think this will work. We should remove this implementation, if we cannot get it right IMO

Reply:
which is indeed hard. I dont its possible on cloud stores.

Reply:
This was just implemented for my testing (like mentioned in the comments), I've moved it to test scope.